spring-kafka-docs/src/main/antora/modules/ROOT/pages/kafka/headers.adoc

```diff
@@ -179,3 +179,80 @@ MessagingMessageConverter converter() {
+
+* **Producer side:** `DefaultKafkaHeaderMapper` writes the JSON bytes, while `SimpleKafkaHeaderMapper` ignore it.
+* **Consumer side:** the mapper exposes the header as a single value—the **last occurrence wins**; earlier duplicates are silently discarded.
+
+For example, on the producer side:
+[source, java]
+----
+Message<String> message = MessageBuilder
+       .withPayload("test-multi-value-header")
+       .setHeader("test-multi-value-header", List.of("test-value1", "test-value2"))
+       .build();
+
+DefaultKafkaHeaderMapper headerMapper = new DefaultKafkaHeaderMapper();
+RecordHeaders mappedHeaders = new RecordHeaders();
+headerMapper.fromHeaders(message.getHeaders(), mappedHeaders);
+
+ObjectMapper objectMapper = new ObjectMapper();
+RecordHeader expectedHeader = new RecordHeader("test-multi-value-header",
+    objectMapper.writeValueAsBytes(List.of("test-value1", "test-value2")));
+assertThat(mappedHeaders.headers(multiValueHeader1).iterator().next()).isEqualTo(expectedHeader);
+----
+
+For example, on the consumer side:
+[source, java]
+----
+RecordHeaders recordHeaders = new RecordHeaders();
+recordHeaders.add("test-multi-value1", new byte[] { 0, 0, 0, 1 });
+recordHeaders.add("test-multi-value1", new byte[] { 0, 0, 0, 2 });
+
+Map<String, Object> mappedHeaders = new HashMap<>();
+headerMapper.toHeaders(recordHeaders, mappedHeaders);
+
+assertThat(headersMapped.get("test-multi-value1")).isEqualTo(new byte[] { 0, 0, 0, 2 });
+----
+
+Preserving each individual header requires explicit registration of patterns that designate the header as multi‑valued.
+
+`DefaultKafkaHeaderMapper#setMultiValueHeaderPatterns(String... patterns)` accepts a list of patterns, which can be either wildcard expressions or exact header names.
+
+[source, java]
+----
+DefaultKafkaHeaderMapper mapper = new DefaultKafkaHeaderMapper();
+
+// Explicit header names
+mapper.setMultiValueHeaderPatterns("test-multi-value1", "test-multi-value2");
+
+// Wildcard patterns for test-multi-value1, test-multi-value2
+mapper.setMultiValueHeaderPatterns("test-multi-*");
+----
+
+Any header whose name matches one of the supplied patterns is
+
+* **Producer side:** written as separate Kafka headers, one per element.
+* **Consumer side:** collected into a `List<?>` that contains the individual header values; each element is returned to the application **after the usual deserialization or type conversion performed by the configured `HeaderMapper`.**
+
+[NOTE]
+====
+Regular expressions are *not* supported; only the `*` wildcard is allowed in simple patterns—supporting direct equality and forms such as:
+
+- `xxx*`
+- `*xxx`
+- `*xxx*`
+- `xxx*yyy`
+====
+
+[IMPORTANT]
+====
+All elements collected under the same multi‑value header **must be of the same Java type**.
+Mixing, for example, `String` and `byte[]` values under a single header key is not supported and will lead to a conversion error.
```

Comment:
I think this is true only if producer side is our `DefaultKafkaHeaderMapper` when each item of the collection header is JSON-seriailized. If producer is some other framework, then we cannot claim this, since there won't be any dedicated type for those items, so everything is going come back to the consumer application as `List<byte[]>`.